drivers: gpio: silabs-siwx91x: Add device runtime support

smalae opened the pull request:
1. Add the port_count field to the configuration structure and allocate different sizes for the ports array based
on whether the node is HP or ULP.
2. Add Device runtime PM for GPIO driver

jhedberg commented:
@smalae for the future, please use titles that give enough context for the entire Zephyr project, as issues & PRs are visible to everyone. I fixed it now for this PR.

asmellby commented on dts/bindings/gpio/silabs,siwx91x-gpio.yaml:
```diff
@@ -11,6 +11,19 @@ properties:
   interrupts:
     required: true
 
+  silabs,port-count:
```
The `gpio-controller` is actually the `silabs,siwx91x-gpio-port`, this is a parent node. That said, it should be possible for the driver to get this using `DT_CHILD_NUM()` instead of it being a property in the binding.

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -183,6 +183,10 @@
 
 			#address-cells = <1>;
 			#size-cells = <0>;
+			silabs,port-count = <4>;
```
Code should use `DT_CHILD_NUM()` instead of this being a property in dts

smalae replied:
Done

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -183,6 +183,10 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
+			silabs,port-count = <4>;
+			silabs,max-pin-usage-count = <8>;
```
Code should use the pin mask constructed from `ngpios` and `gpio-reserved-ranges`, which is already available as the `.common.port_pin_mask` member of the config struct on each port. No need for another dts property.

asmellby commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -158,6 +135,47 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+		pdata->pin_config_info[pdata->cur_pin].port_dev = dev;
+		pdata->pin_config_info[pdata->cur_pin].pin = pin;
+		pdata->pin_config_info[pdata->cur_pin].flags = flags;
+		pdata->cur_pin++;
```
A pin can be configured multiple times, this needs to reuse an existing allocation if the same pin is reconfigured.

smalae replied:
Done

smalae committed:
dts: arm: silabs-siwx91x: Update GPIO nodes

Add soc power domain and zephyr,pm-device-runtime-auto

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: gpio: silabs-siwx91x: Differentiate port count

Add the port_count field to the configuration structure
and allocate different sizes for the ports array based
on whether the node is HP or ULP.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: gpio: silabs-siwx91x: Device runtime PM

This commit enables the pm device runtime support
for the siwx91x gpio driver.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: gpio: silabs-siwx91x: clang format corrections

Apply clang format corrections

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

jhedberg commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -158,6 +136,67 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+			port_cfg = port_dev[i]->config;
+			port_data = port_dev[i]->data;
+			pin_cnt = 0;
+			while (port_data->pin_config_info[i].port_dev != NULL &&
+			       pin_cnt < __builtin_popcount(port_cfg->common.port_pin_mask)) {
+				ret = gpio_siwx91x_pin_configure(
+					port_data->pin_config_info[pin_cnt].port_dev,
+					port_data->pin_config_info[pin_cnt].pin,
+					port_data->pin_config_info[pin_cnt].flags);
```
Is it really correct that you're using a different index for `port_data->pin_config_info` in the while loop condition and inside the loop? Looks a bit suspicious.

jhedberg commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -158,6 +136,67 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+
+static int gpio_siwx91x_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct gpio_siwx91x_common_config *pcfg = dev->config;
+	struct gpio_siwx91x_common_data *pdata = dev->data;
```
Does the `p` prefix for these variable names stand for "port", or are you using Hungarian notation? In both cases I think you should just drop it.

smalae committed:
drivers: gpio: silabs-siwx91x: Refactor config structures

Drop usage of `pcfg` and `pdata` in favor of unified `cfg`
for common configuration and `port_cfg`, `port_data` for
port-specific configurations.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

Martinhoff-maker commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -58,59 +64,32 @@ struct gpio_siwx91x_port_data {
-	uint16_t pin_direction[MAX_PIN_COUNT];
-#endif
+	struct gpio_siwx91x_pin_config_info *pin_config_info;
+	uint8_t next_pin;
```
```next_pin``` feels not the good name for this variable. If I understand right, it saves the number of configured pin in the port ? would rather use "pin_ctn" or "pin_count" to have link with port variable.

Martinhoff-maker commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -127,86 +106,147 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+		port_data->next_pin++;
+	}
 
+	if (cur_cfg_pin < __builtin_popcount(port_cfg->common.port_pin_mask)) {
```
This value is already known and constant, it is ARRAY_SIZE(port_data->pin_config_info)

Martinhoff-maker commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -127,86 +106,147 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+		port_data->pin_config_info[cur_cfg_pin].pin = pin;
+		port_data->pin_config_info[cur_cfg_pin].flags = flags;
+	} else {
+		return -EIO;
```
-EINVAL feels to be a better return value.
-EIO is for: "I/O error when accessing an external GPIO chip."

Martinhoff-maker commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -127,86 +106,147 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+			port_cfg = port_dev[i]->config;
+			port_data = port_dev[i]->data;
+			pin_cnt = 0;
+			while (port_data->pin_config_info[pin_cnt].port_dev != NULL &&
+			       pin_cnt < __builtin_popcount(port_cfg->common.port_pin_mask)) {
+				ret = gpio_siwx91x_pin_configure(
+					port_data->pin_config_info[pin_cnt].port_dev,
+					port_data->pin_config_info[pin_cnt].pin,
+					port_data->pin_config_info[pin_cnt].flags);
+				if (ret) {
+					return ret;
+				}
+				pin_cnt++;
+			}
```
can't you use the fact that you knows how much pin has been configured with your ```next_pin``` variable in port data ?

```suggestion
			for(int i = 0; i < port_data->next_pin; i++) {
				ret = gpio_siwx91x_pin_configure(
					port_data->pin_config_info[i].port_dev,
					port_data->pin_config_info[i].pin,
					port_data->pin_config_info[i].flags);
				if (ret) {
					return ret;
				}
			}
```

or something like this ? renaming to "pin_ctn" make more sense too here.

Martinhoff-maker commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -127,86 +106,147 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
+					port_data->pin_config_info[pin_cnt].pin,
+					port_data->pin_config_info[pin_cnt].flags);
+				if (ret) {
+					return ret;
```
maybe a stupid question question but how could it failed since you already managed to configure this pin this way ? and even if it failed, what our power domain driver will do about that info ? (more open question)

smalae commented:
> @smalae any update on this?

"I’m currently testing this while juggling a few other tasks. I’ll make sure to address it soon

smalae commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -58,48 +64,21 @@ struct gpio_siwx91x_port_data {
 	const struct gpio_siwx91x_common_config *pcfg = parent->config;
+	uint8_t cur_cfg_pin = 0;
 	sl_status_t status;
+	int i;
```
Updated
